src/main/scala/io/iohk/ethereum/network/PeersInfoHolderActor.scala

```diff
@@ -0,0 +1,189 @@
+import io.iohk.ethereum.db.storage.AppStateStorage
+import io.iohk.ethereum.network.PeerActor.{DisconnectPeer, SendMessage}
+import io.iohk.ethereum.network.PeersInfoHolderActor._
+import io.iohk.ethereum.network.PeerEventBusActor.{PeerSelector, Subscribe, Unsubscribe}
+import io.iohk.ethereum.network.PeerEventBusActor.PeerEvent._
+import io.iohk.ethereum.network.PeerEventBusActor.SubscriptionClassifier._
+import io.iohk.ethereum.network.handshaker.Handshaker.HandshakeResult
+import io.iohk.ethereum.network.p2p.Message
+import io.iohk.ethereum.network.p2p.messages.CommonMessages.{NewBlock, Status}
+import io.iohk.ethereum.network.p2p.messages.PV62.{BlockHeaders, GetBlockHeaders, NewBlockHashes}
+import io.iohk.ethereum.network.p2p.messages.WireProtocol.Disconnect
+
+class PeersInfoHolderActor(peerEventBusActor: ActorRef, appStateStorage: AppStateStorage,
+                           forkResolverOpt: Option[ForkResolver]) extends Actor with ActorLogging {
+
+  val msgCodesWithInfo: Set[Int] = Set(BlockHeaders.code, NewBlock.code, NewBlockHashes.code)
+
+  peerEventBusActor ! Subscribe(PeerHandshaked)
+
+  override def receive: Receive = handlePeersInfoEvents(Map.empty)
+
+  def handlePeersInfoEvents(peersWithInfo: Map[PeerId, PeerWithInfo]): Receive = {
+
+    case MessageToPeer(message, peerId) if peersWithInfo.contains(peerId) =>
+      val PeerWithInfo(peer, oldPeerInfo) = peersWithInfo(peerId)
+      val newPeerInfo = handleSentMessage(message, oldPeerInfo)
+      context become handlePeersInfoEvents(peersWithInfo + (peerId -> PeerWithInfo(peer, newPeerInfo)))
+
+    case MessageFromPeer(message, peerId) if peersWithInfo.contains(peerId) =>
+      val PeerWithInfo(peer, oldPeerInfo) = peersWithInfo(peerId)
+      val newPeerInfo = handleReceivedMessage(message, oldPeerInfo, peer)
+      context become handlePeersInfoEvents(peersWithInfo + (peerId -> PeerWithInfo(peer, newPeerInfo)))
+
+    case PeerHandshakeSuccessful(peer, peerInfo: PeerInfo) =>
+      peerEventBusActor ! Subscribe(PeerDisconnectedClassifier(PeerSelector.WithId(peer.id)))
+      peerEventBusActor ! Subscribe(MessageReceivedClassifier(msgCodesWithInfo, PeerSelector.WithId(peer.id)))
+      peerEventBusActor ! Subscribe(MessageSentClassifier(msgCodesWithInfo, PeerSelector.WithId(peer.id)))
+
+      //Ask for the highest block from the peer
+      peer.ref ! SendMessage(GetBlockHeaders(Right(peerInfo.remoteStatus.bestHash), 1, 0, false))
+      context become handlePeersInfoEvents(peersWithInfo + (peer.id -> PeerWithInfo(peer, peerInfo)))
+
+    case PeerDisconnected(peerId) if peersWithInfo.contains(peerId) =>
+      peerEventBusActor ! Unsubscribe(PeerDisconnectedClassifier(PeerSelector.WithId(peerId)))
+      peerEventBusActor ! Unsubscribe(MessageReceivedClassifier(msgCodesWithInfo, PeerSelector.WithId(peerId)))
+      peerEventBusActor ! Unsubscribe(MessageSentClassifier(msgCodesWithInfo, PeerSelector.WithId(peerId)))
```

Comment:
They are separate as the `MessageReceivedClassifier` is for messages received from peers while `MessageSentClassifier` is for messages sent to them.
Apart from the `PeersInfoHolder`, currently no other actor is interested in both kinds of events as for example the `SyncBlockHeadersRequestHandler` is only interested in receiving the `BlockHeaders` that it requested. This is also the case for the `PendingTransactionsManager` and all the `SyncRequestHandler`s.